File: term-frequency.go
package main

import (
	"fmt"
	"sort"
)

type Word struct {
	Name  string
	Count int32
}

// Implement the sorting
type Words []*Word

func (words Words) Len() int {
	return len(words)
}
func (words Words) Less(i, j int) bool {
	if words[i].Count != words[j].Count {
		return words[i].Count > words[j].Count
	}
	return words[i].Name < words[j].Name
}
func (words Words) Swap(i, j int) {
	words[i], words[j] = words[j], words[i]
}

func main() {
	// Load words
	wordLength := 0
	fmt.Scanf("%d", &wordLength)
	rawWords := make([]string, wordLength)
	for i := 0; i < wordLength; i++ {
		fmt.Scanf("%s", &rawWords[i])
	}

	// Count the frequencies
	frequencies := map[string]int32{}
	for _, word := range rawWords {
		if _, ok := frequencies[word]; ok {
			frequencies[word] += 1
		} else {
			frequencies[word] = 1
		}
	}

	// Sort words
	words := Words{}
	for wordName, count := range frequencies {
		words = append(words, &Word{wordName, count})
	}
	sort.Sort(words)

	// Print words
	outputWordLength := 0
	fmt.Scanf("%d", &outputWordLength)
	for i := 0; i < outputWordLength; i++ {
		fmt.Println(words[i].Name)
	}
}

File: term-frequency.rb
# Load inputs
word_length = gets.to_i
words = []
word_length.times.each{ words << gets.strip }
output_word_length = gets.to_i

## lexicographical

# Count word frequencies
# Example value of word_frequencies:
#   {"Fee": 12, "Fum": 2}
word_frequencies = {}
words.each do |word|
  word_frequencies[word] ||= 0
  word_frequencies[word] += 1
end


WORD = 0
FREQUENCY = 1
sorted_word_frequencies = word_frequencies.to_a.sort do |a, b|
  f = b[FREQUENCY] <=> a[FREQUENCY]
  f != 0 ? f : a[WORD] <=> b[WORD]
end

sorted_word_frequencies.first(output_word_length).each{|wf| puts wf[WORD]}
